tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc

```diff
@@ -0,0 +1,369 @@
+namespace gpu {
+namespace {
+namespace m = match;
+
+bool IsFMHACustomCall(const HloInstruction* instr) {
+  return instr->opcode() == HloOpcode::kCustomCall &&
+         (instr->custom_call_target() == kCudnnfMHABmmBmmCallTarget ||
+          instr->custom_call_target() ==
+              kCudnnfMHAScaleBiasMaskSoftmaxCallTarget ||
+          instr->custom_call_target() ==
+              kCudnnfMHAScaleBiasMaskSoftmaxDropoutCallTarget ||
+          instr->custom_call_target() == kCudnnfMHAScaleMaskSoftmaxCallTarget ||
+          instr->custom_call_target() ==
+              kCudnnfMHAScaleMaskSoftmaxDropoutCallTarget ||
+          instr->custom_call_target() == kCudnnfMHASoftmaxDropoutCallTarget);
+}
+
+StatusOr<bool> IsBatchedDot(const HloInstruction* gemm) {
+  GemmBackendConfig config;
+  TF_ASSIGN_OR_RETURN(config, gemm->backend_config<GemmBackendConfig>());
+  const DotDimensionNumbers& dot_dims = config.dot_dimension_numbers();
+  bool is_batch_dot = !dot_dims.lhs_batch_dimensions().empty() ||
+                      !dot_dims.rhs_batch_dimensions().empty();
+  return is_batch_dot;
+}
+
+bool IsBatchedMatmul(const HloInstruction* gemm) {
+  return IsCublasGemm(*gemm) && IsBatchedDot(gemm).value();
+}
+
+// Give this instruction a more useful name than "custom-call.42".
+Status SetName(HloModule* module, HloInstruction* fmha) {
+  if (fmha->custom_call_target() == kCudnnfMHABmmBmmCallTarget) {
+    module->SetAndUniquifyInstrName(fmha, "fmha-bmm-bmm");
+    return OkStatus();
+  }
+  if (fmha->custom_call_target() == kCudnnfMHASoftmaxDropoutCallTarget) {
+    module->SetAndUniquifyInstrName(fmha, "fmha-bmm-softmax-dropout-bmm");
+    return OkStatus();
+  }
+  if (fmha->custom_call_target() == kCudnnfMHAScaleMaskSoftmaxCallTarget) {
+    module->SetAndUniquifyInstrName(fmha, "fmha-bmm-scale-mask-softmax-bmm");
+    return OkStatus();
+  }
+  if (fmha->custom_call_target() ==
+      kCudnnfMHAScaleMaskSoftmaxDropoutCallTarget) {
+    module->SetAndUniquifyInstrName(fmha,
+                                    "fmha-bmm-scale-mask-softmax-dropout-bmm");
+    return OkStatus();
+  }
+  if (fmha->custom_call_target() == kCudnnfMHAScaleBiasMaskSoftmaxCallTarget) {
+    module->SetAndUniquifyInstrName(fmha,
+                                    "fmha-bmm-scale-bias-mask-softmax-bmm");
+    return OkStatus();
+  }
+  if (fmha->custom_call_target() ==
+      kCudnnfMHAScaleBiasMaskSoftmaxDropoutCallTarget) {
+    module->SetAndUniquifyInstrName(
+        fmha, "fmha-bmm-scale-bias-mask-softmax-dropout-bmm");
+    return OkStatus();
+  }
+
+  return InternalError(
+      "Found invalid FMHA custom-call target while setting custom-call name");
+}
+
+bool IsComputeCapabilitySupported(stream_executor::CudaComputeCapability cc) {
+  if (!(cc.IsAtLeast(
+          se::CudaComputeCapability::AMPERE)  && cc.minor == 0 )) {
+      VLOG(2) << "CudnnFusedMHARewriter did not run. Unsupported compute "
+                 "capability. Only GA100 supported.";
```

Comment:
Nit: it's not GA100, it's compute capability? Should it also check for cuDNN version?

Reply:
It's actually GA100. Hence checking the minor as well. I will add code to the cudnn version in the next commit.